Make this a JavaScript library

ivteplo opened the pull request:
Related issue: #1 

Things to do:
- [x] Create a `BottomSheet` class
- [x] Force the developer to specify an identifier for the sheet
- [x] Set up the `aria-controls` property for a button that closes a sheet
- [x] Set up the `aria-controls` property for a button that opens the sheet (just to make the example code better)
- [ ] Find a way to limit access of screen reader and keyboard to the contents behind the sheet
- [x] Find a better way to generate the DOM
- [x] Rename CSS classes
- [x] Find a way to scope styles (so that they won't break user's layout)
- [x] Add more options to customize the sheet
- [x] Maybe create a custom tag as another way to use the component (?)
- [x] Implement custom events (hide/show)
- [x] Add a method to unmount the component and remove its event listeners from the `window` object
- [ ] ~~Add the title area in the top bar of the sheet~~ - This can be a future enhancement
- [x] Write library documentation

ivteplo committed:
Create a `BottomSheet` class

n-ce commented:
Hello, @ivteplo , I think you should drop the legendary ultimate DOM selector as it is a costly operation compared to just getElementById. It allows just as much readability but also the highest performance which is essential for a smooth ux.

ivteplo committed:
Set up Vite for building and testing

ivteplo committed:
Work in process: add basic functionality, use CSS modules

ivteplo committed:
Improve DOM generation code by using JSX

ivteplo committed:
Update utilities.css version in the example code

ivteplo committed:
Force the developer to specify an identifier for the sheet

ivteplo committed:
Support `reference` in the `createElement` attribute to remove cryptic `querySelector`s

ivteplo commented:
Maybe this will be useful to someone who is also interested in how to 'trap focus' inside of a modal:
https://css-tricks.com/a-css-approach-to-trap-focus-inside-of-an-element/

ivteplo committed:
Update styles. Don't use CSS modules

ivteplo committed:
Rename CSS classes

ivteplo committed:
Add a method to unmount the bottom sheet component / restore initial HTML

ivteplo committed:
Fix test (`aria-controls` wasn't specified in HTML). Improve API

Now API is more universal. The sheet body won't be automatically replaced with the component code.
Instead, the library user should do it personally.

`aria-controls` was set to `""` in HTML.

The test has been updated according to the API changes.
Additionally, `<template>` is now used to store sheet body
(so that it won't be shown to the user in case the page will be loading for a long time)

ivteplo committed:
Merge branch 'main' of https://github.com/ivteplo/bottom-sheet into library

ivteplo committed:
Add some documentation

ivteplo committed:
Improve example inside README.md

ivteplo committed:
Use null-coalescing operator when calling reference setter in `createElement`

ivteplo commented:
As far as I can see, focus trapping is quite a problematic feature. Probably, it would be better to let the user implement it instead (e.g. using a special library).

Why I consider this feature problematic to implement here:
- Implementing focus trapping here myself would be somewhat excessive. Additionally, I would have to do a lot of testing.
- Adding a direct dependency to a focus-trapping library might be a problem for people using the same library of a different version, since both versions would get included in the bundle and some incompatibility bugs might arise (especially when stacking multiple sheets or dialogs). 
- Adding a peer dependency to a focus-trapping library would force people follow my choice or, again, compatibility issues might arise.

I might create a tutorial on how to trap the focus inside the sheet with a separately-installed focus-trapping library (or libraries).

ivteplo committed:
Fix wrong library path in the test page

ivteplo committed:
Trigger `hide` and `show` events on the sheet

ivteplo committed:
Document the new code. Improve names and code style

ivteplo committed:
Update documentation. Don't minify the build output

ivteplo committed:
Switch to HTML Custom Elements API. Stop changing height

Fixes #3: the height of the sheet won't change anymore. Instead, the `translateY` and `scale` get changed.

ivteplo committed:
Adapt styles for desktop and tablets

ivteplo committed:
Dispatch custom events

ivteplo committed:
Add more comments

ivteplo committed:
Update folder structure

ivteplo committed:
Make the library user define the element in the registry

Thus the user can rename the component and will not have to solve conflicts with other components

ivteplo committed:
Fix the sheet not closing on backdrop click

ivteplo committed:
Fix the error that prevented the dragging process

ivteplo committed:
Generate simple API documentation

ivteplo committed:
Improve documentation generation

ivteplo committed:
Bind options to HTML attributes

ivteplo committed:
Update documentation
